rabbitmq/retryable_consumer.go

```diff
@@ -0,0 +1,123 @@
+// Copyright 2019 SumUp Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rabbitmq
+
+import (
+	"context"
+	"time"
+
+	"github.com/palantir/stacktrace"
+	"go.uber.org/zap"
+
+	"github.com/sumup-oss/go-pkgs/backoff"
+	"github.com/sumup-oss/go-pkgs/logger"
+)
+
+type RetryableConsumer struct {
+	config        RetryableConsumerConfig
+	logger        logger.StructuredLogger
+	metric        Metric
+	handler       Handler
+	clientFactory func(ctx context.Context, config *ClientConfig) (*RabbitMQClient, error)
+}
+
+type RetryableConsumerConfig struct {
+	MaxRetryAttempts int
+	// healthCheckFactor is a number representing how much N multiplied by backoffConfig.Max time is needed
+	// for a block of code to run w/o returning an error, to consider it healthy.
+	// E.g backConfig.Max = 1min, healthCheckFactor = 2, means that code needs to run 2min at least to be healthy
+	// and retried again starting from backoffConfig.Base the next time it has an error.
+	HealthCheckFactor  int
+	BackoffConfig      *backoff.Config
+	ConsumerConfig     ConsumerConfig
+	RabbitClientConfig *ClientConfig
+}
+
+func NewRetryableConsumer(
+	newClientFactory func(ctx context.Context, config *ClientConfig) (*RabbitMQClient, error),
+	config RetryableConsumerConfig,
+	logger logger.StructuredLogger,
+	metric Metric,
+	handler Handler,
+) *RetryableConsumer {
+	return &RetryableConsumer{
+		clientFactory: newClientFactory,
+		config:        config,
+		handler:       handler,
+		logger:        logger,
+		metric:        metric,
+	}
+}
+
+func (c *RetryableConsumer) Run(ctx context.Context) error {
+	consumerBackoff := backoff.NewBackoff(c.config.BackoffConfig)
+	currentRetryAttempts := 0
+
+	for {
+		startTime := time.Now()
+		err := c.doRun(ctx)
+		if err != nil {
```

Comment:
Are we absolutely sure there is no use case for the consumer to return `nil` here that is not caused by `<-ctx.Done()` being handled somewhere inside the consumer. Would it be safer if we remove the `return nil` at the end of this for loop and instead start the loop with an explicit check on `<-ctx.Done()`, i.e.
```
for {
	select {
		case <-ctx.Done():
			return nil
		default:
			// the current implementation
	}
}
```

Reply:
We discussed this with @syndbg  and @perkon. In the doRun receiver func you have 
`	if ctx.Err() != nil {
		c.logger.Info("received context cancel")
		return nil
	}
`
That will always check for the context.